sljit_src/sljitWXExecAllocator.c

```diff
@@ -0,0 +1,77 @@
+/*
+ *    Stack-less Just-In-Time compiler
+ *
+ *    Copyright Zoltan Herczeg (hzmester@freemail.hu). All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are
+ * permitted provided that the following conditions are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright notice, this list of
+ *      conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form must reproduce the above copyright notice, this list
+ *      of conditions and the following disclaimer in the documentation and/or other materials
+ *      provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) AND CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER(S) OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+   This file contains a simple W^X executable memory allocator for POSIX systems.
+
+   It allocates a separate block for each code block which may waste a lot of memory.
+*/
+
+SLJIT_API_FUNC_ATTRIBUTE void* sljit_malloc_exec(sljit_uw size)
+{
+	sljit_uw* ptr;
+
+	size += sizeof(sljit_uw);
+	ptr = (sljit_uw*)mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+
+	if (ptr == NULL) {
+		return NULL;
+	}
+
+	*ptr = size;
+	return ptr + 1;
+}
+
+SLJIT_API_FUNC_ATTRIBUTE void sljit_free_exec(void* ptr)
+{
+	sljit_uw *start_ptr = ((sljit_uw*)ptr) - 1;
+	munmap(start_ptr, *start_ptr);
+}
+
+#define SLJIT_UPDATE_WX_FLAGS(from, to, enable_exec) \
+	sljit_update_wx_flags((from), (to), (enable_exec))
+
+static void sljit_update_wx_flags(void *from, void *to, sljit_s32 enable_exec)
+{
+	static sljit_uw page_mask = 0;
+
+	sljit_uw addr = (sljit_uw)from;
+	int prot = PROT_READ | (enable_exec ? PROT_EXEC : PROT_WRITE);
+
+	if (page_mask == 0) {
+		/* This is thread safe, since the value is constant. */
+		page_mask = (sljit_uw)(sysconf(_SC_PAGESIZE) - 1);
+	}
+
+	addr &= ~page_mask;
+
+	mprotect((void*)addr, (sljit_uw)to - addr, prot);
```

Comment:
.. or casting the pointers to int

for this allocator, size should be always be the requested size anyway, but if some other allocator try to use this interface (and the system allows it) they might only switch part of the whole allocation in pagesize multiples

Reply:
I may missing something in your comment but the point is switching only a part of the allocated range.

Reply:
mprotect works (and this is architecture/system specific) in pages at a minimum and old versions of it used to throw errors when passing sizes that were not multiples of the system pagesize to make the point clear.

in my implementation (which you copied) I mostly ignored that (since it works in NetBSD anyway, at least in amd64) but I was expecting (and getting) pagesize rounded values from mmap which I also didn't track exactly since munmap/mprotect didn't care either, and because @krytarowski was going to have to fix it when doing the same for bpfjit (in the NetBSD kernel)

this allocator SHOULD switch all the pages allocated between modes, and because if the size is bigger than a page then it will target the wrong page resulting in a segfault.

other allocators that might be using a macro like this might instead try to only change the page that they care for which could be smaller than the allocated size.

the complexities on calculating the end and start (as well as avoiding UB or using ints to do pointer subtraction) is why I suggested instead to pass a size_t even if it would be a different interface than cache clearing.

Reply:
I confirm that rounding to PAGE_SIZE is wanted. It is abstracted in userland, but in kernel we are stricter.

Reply:
@krytarowski the optional rounding available as part of followup PR #63